src/main/java/teammates/ui/controller/InstructorCourseInstructorAddAction.java

```diff
@@ -162,73 +197,158 @@ private void updateInstructorWithSectionLevelPrivileges(String courseId, Instruc
+                                         i, sectionNameFromParam);
                 }
             }
         }
-        return sectionNamesMap;
     }
-
-    private void updateInstructorPrivilegesForSectionInSectionLevel(String sectionParam, List<String> sectionNames,
-                                                                    InstructorAttributes instructorToAdd) {
+    
+    /**
+     * Marks {@code sectionToMark} as special in the associated mappings.
+     * 
+     * @param isSectionSpecialMappings       Mapping of names of sections to boolean values indicating if they are special.
+     *                                           This will be modified within the method.
+     * @param specialSectionsInSectionGroups Mapping of section group names to the special sections that they contain.
+     *                                           This will be modified within the method.
+     * @param sectionGroupIndex              Index of the section group to be updated.
+     * @param sectionToMark                  Section that will be marked as special.
+     */
+    private void markSectionAsSpecial(HashMap<String, Boolean> isSectionSpecialMappings,
+            HashMap<String, List<String>> specialSectionsInSectionGroups, int sectionGroupIndex,
+            String sectionToMark) {
+        // indicate that section group covers the section
+        // and mark that this section is special
+        String sectionGroupParamName = Const.ParamsNames.INSTRUCTOR_SECTION_GROUP + sectionGroupIndex;
+        if (specialSectionsInSectionGroups.get(sectionGroupParamName) == null) {
+            specialSectionsInSectionGroups.put(sectionGroupParamName, new ArrayList<String>());
+        }
+        specialSectionsInSectionGroups.get(sectionGroupParamName).add(sectionToMark);
+        isSectionSpecialMappings.put(sectionToMark, true);
+    }
+    
+    /**
+     * Updates instructor privileges at section level by retrieving request parameters.
+     * The parameters that are retrieved are based off {@code sectionGroupName}.
+     * 
+     * @param sectionGroupName              Name of the section group.
+     * @param specialSectionsInSectionGroup Sections marked as special under the section group.
+     * @param instructorToAdd               Instructor that will be added.
+     *                                          This will be modified within the method.
+     */
+    private void updateInstructorPrivilegesForSectionInSectionLevel(String sectionGroupName,
+            List<String> specialSectionsInSectionGroup, InstructorAttributes instructorToAdd) {
         boolean isViewStudentInSectionsChecked =
-                getRequestParamValue(Const.ParamsNames.INSTRUCTOR_PERMISSION_VIEW_STUDENT_IN_SECTIONS + sectionParam) != null;
+                getRequestParamValue(Const.ParamsNames.INSTRUCTOR_PERMISSION_VIEW_STUDENT_IN_SECTIONS
+                        + sectionGroupName) != null;
```

Comment:
I don't usually pick on indentations and stuff, but that `+` would be better under the `[C]onst`.